Convert pik to cik

qnhn22 opened the pull request:
# Description

I have created a new state called 'Cik' for client-side usage. The code has been refactored to eliminate the need for the 'pik' array, replacing it with the newly introduced 'Cik' state. In order to make 'Cik' accessible, I have developed a new API that fetches 'Cik' data, which includes updating the model, creating an SQL file for retrieving information from the database, and implementing a new route.  
  
Additionally, since the 'pik' is no longer required, the unit index can be removed from the unit table and refactor the code to eliminate any usage of the unit index. (database changed)


Fixes #901

(In general, OED likes to have at least one issue associated with each pull request. Replace [issue] with the OED GitHub issue number. In the preview you will see an issue description if you hover over that number. You can create one yourself before doing this pull request. This is where details are normally given on what is being addressed. Note you should not use the word "Fixes" if it does not completely address the issue since the issue would automatically be closed on merging the pull request. In that case use "Partly Addresses #[issue].)

## Type of change

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [x] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [ ] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.github.io/developer/cla.html) and each author is listed in the Description section.

## Limitations

(Describe any issues that remain or work that should still be done.)

qnhn22 committed:
create Cik model, route, api, and state

qnhn22 committed:
refactor front-end based on the Cik

qnhn22 committed:
refactor front-end based on the Cik

qnhn22 committed:
update comments

qnhn22 committed:
Merge branch 'development' into convert_pik_to_cik

qnhn22 committed:
add header

huss commented on src/client/app/components/groups/EditGroupModalComponent.tsx:
```diff
@@ -355,7 +356,7 @@ export default function EditGroupModalComponent(props: EditGroupModalComponentPr
 		// metersState normally does not change but can so include.
 		// globalGroupsState can change if another group is created/edited and this can change ones displayed in menus.
-	}, [ConversionArray.pikAvailable(), metersState, globalGroupsState, groupState.defaultGraphicUnit, groupState.deepMeters]);
+	}, [metersState, globalGroupsState, groupState.defaultGraphicUnit, groupState.deepMeters]);
```
Other places have ciksState when ConversionArray.pikAvailable() removed. Should it be here?

qnhn22 replied:
Actually no. Because the Cik data is fetched in the initialization, ciksState shouldn't be here. Thank you for your suggestion.

huss replied:
A couple of thoughts on this:

1. It is done in createGroupModalComponent which is very similar so I'm unclear on why not here.
2. The assumption where is that the initialization must be completed before the admin gets to this modal. I agree that is quite likely but is not guaranteed. I believe it happened with pik during a manual refresh of the group page. I'm not positive it can happen but I think it has so it seems safer to include the condition, esp. as this is very similar to pik that needed it.

qnhn22 replied:
I just thought that the initialization step run first, then it would make Cik available before accessing the modals or dropdown which use the Cik data. But as you suggested, I will include it as a condition because it would be safer. Thank you.

huss commented on src/client/app/components/meters/CreateMeterModalComponent.tsx:
```diff
@@ -122,6 +121,7 @@ export default function CreateMeterModalComponent(props: CreateMeterModalCompone
 
 	// Dropdowns
 	const [dropdownsState, setDropdownsState] = useState(dropdownsStateDefaults);
+	const globalCiksState = useSelector((state: State) => state.ciks.ciks);
```
In editMeterModalComponent this is ciksState. Should the name be consistent?

huss commented on src/client/app/types/redux/ciks.ts:
```diff
@@ -0,0 +1,39 @@
+	ciks: CikData[];
+}
+
+
```
Please remove all but one last blank line.

huss commented on src/client/app/types/conversionArray.ts:
```diff
@@ -1,28 +1,28 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
-* License, v. 2.0. If a copy of the MPL was not distributed with this
-* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// /* This Source Code Form is subject to the terms of the Mozilla Public
```
Given pik is now gone from the code, maybe this file should be removed rather than commented out? There is a record in git if it is ever needed.

huss commented on src/server/models/Cik.js:
```diff
@@ -20,6 +33,25 @@ class Cik {
+	/**
+	 * Create a new Cik object from row's data.
+	 * @param {*} row The row from which Cik will be created.
+	 * @returns {Promise.<Array.<Cik>>}
```
This seems to return an object and not a promise. Is it a promise?

huss commented on src/server/routes/ciks.js:
```diff
@@ -0,0 +1,36 @@
+const { log } = require('../log');
+const { getConnection } = require('../db');
+const Cik = require('../models/Cik');
+const { success, failure } = require('./response');
```
I don't see where these two imports or the next one is used.

huss commented on src/server/routes/conversions.js:
```diff
@@ -1,6 +1,6 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
-  * License, v. 2.0. If a copy of the MPL was not distributed with this
-  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+	* License, v. 2.0. If a copy of the MPL was not distributed with this
```
I think OED normally uses a single space so the * linrd up on all lines.

huss reviewed (changes requested):
Thanks to @qnhn22 for working on this issue. Testing indicates it works as expected. I have a few modest comments to address. Also, I noticed that a search reveals places where "pik" and "conversionArray" still appear (or are part of a word) in the code or comments. I looked at several of them and it is not clear to me that they are needed. Do you want to check them and remove as possible or is that something I need to do?

qnhn22 committed:
reset the Cik state when adding or editing conversion

qnhn22 committed:
fix wrong file path

qnhn22 committed:
remove pik

huss committed:
purge unit_index/unitIndex

This removes from DB, renames to better name for source/destination, purges from code.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/qnhn22/1109

huss commented:
I just pushed changes to this PR. They do the final removal for the unit index from code and DB. It also merged in development to be sure there are no issues. I noted in the description that this changes the DB configuration and announced on Discord. With @qnhn22 changes and these, this PR is hopefully ready to merge. This completes the design doc steps until "Integrating cik".
